### PR TITLE
Move ConsistentListFromCache to Beta default

### PR DESCRIPTION
**Pull request by serathius**
/kind feature

Ref https://github.com/kubernetes/enhancements/issues/2340
```release-note
Move ConsistentListFromCache feature flag to Beta and enable it by default
```

### CODE REVIEW

**Comment by serathius** [comment]
/retest

Don't have any idea how to debug `test-cmd: run_recursive_resources_tests` tests.

cc @dead2k

**Comment by serathius** [comment]
Managed to debug the test, look like we ConsistentListFromCache allowed serving non-recursive lists with RV="" from cache, but didn't request progress notification as I didn't think we need to add it to `WaitUntilFreshAndGet` function.

**Comment by serathius** [comment]
We got a pass! Unfortunately I found that the integration tests are flaking due to bug in etcd. Will link the bug soon

:( 

/test pull-kubernetes-integration

**Comment by serathius** [comment]
With etcd v3.5.13 the tests are green!

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -1250,7 +1250,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	genericfeatures.APIServingWithRoutine: {Default: true, PreRelease: featuregate.Beta},
 
-	genericfeatures.ConsistentListFromCache: {Default: false, PreRelease: featuregate.Alpha},
+	genericfeatures.ConsistentListFromCache: {Default: true, PreRelease: featuregate.Beta},
```
Comment (liggitt) [inline pkg/features/kube_features.go:1253]: what is the behavior of this when we aren't running against the minimum etcd version we expect?

Comment (serathius) [reply pkg/features/kube_features.go:1253]: Ups, this is still a draft waiting for https://github.com/kubernetes/kubernetes/pull/123264 to be merged and ensure minimal etcd version.

**Comment by serathius** [comment]
/cc @wojtek-t 

We are almost ready, just need https://github.com/kubernetes/kubernetes/pull/123264

**Comment by wojtek-t** [comment]
I'm waiting with approving this PR to have a "consistency checker" implemented and enabled in CI.
@p0lyn0mial is working on one for streaming-lists and we should be able to reuse the same thing

**Comment by dims** [comment]
cc @shyamjvs

**Comment by serathius** [comment]
Second time the test failed, think its irrelevant, will need to rebase.
```
=== FAIL: k8s.io/kubectl/pkg/cmd/logs TestLog/stateful_set_logs_with_all_pods (0.05s)
    logs_test.go:557: stateful set logs with all pods: expected to contain "[pod/test-sts-0/test-container] test log content for pod test-sts-0\n[pod/test-sts-1/test-container] test log content for pod test-sts-1\n". Output: "[pod/test-sts-1/test-container] test log content for pod test-sts-1\n[pod/test-sts-0/test-container] test log content for pod test-sts-0\n"
```

**Comment by p0lyn0mial** [comment]
Before enabling this feature shouldn't we change the cacher to honour the Limit for rows `1` and `4` described in [the KEP](https://github.com/kubernetes/enhancements/blob/9f92967516301bbd564b7db221fe92ce3a253792/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#what-if-the-watch-cache-is-stale) ?

**Comment by wojtek-t** [comment]
> Before enabling this feature shouldn't we change the cacher to honour the Limit for rows 1 and 4 described in [the KEP](https://github.com/kubernetes/enhancements/blob/9f92967516301bbd564b7db221fe92ce3a253792/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#what-if-the-watch-cache-is-stale) ?

Good point - we should.
My main question is why this didn't bubble via any test failures...

**Comment by p0lyn0mial** [comment]
> My main question is why this didn't bubble via any test failures...

It turns out that all requests are directed to etcd because the cache doesn't honor the limit parameter.

**Comment by serathius** [comment]
Waiting for https://github.com/kubernetes/kubernetes/pull/125405

**Comment by serathius** [comment]
Oh 
image
cc @wojtek-t

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 2]
```diff
@@ -392,7 +393,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	WatchList: {Default: false, PreRelease: featuregate.Alpha},
 
-	ConsistentListFromCache: {Default: false, PreRelease: featuregate.Alpha},
+	ConsistentListFromCache: {Default: true, PreRelease: featuregate.Beta},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:396]: Unit-test failure seems to be lack of:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L658-L665

in cacher

@p0lyn0mial ^^

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/features/kube_features.go:396]: Why we need more testing of etcd storage.

cc @logicalhan @jpbetz @deads2k

**Comment by wojtek-t** [comment]
hmm - wondering why I didn't see that yet elsewhere

in general https://github.com/kubernetes/kubernetes/pull/125483 once merged should solve it [watchcache initialization will become part of /healthz which integration tests already block on], but that's not yet ready

**Comment by serathius** [comment]
Send a fix in https://github.com/kubernetes/kubernetes/pull/125586

**Comment by serathius** [comment]
We are back to CI being green! Waiting for https://github.com/kubernetes/kubernetes/pull/125405

**Comment by wojtek-t** [comment]
> It turns out that all requests are directed to etcd because the cache doesn't honor the limit parameter.

This effectively means that it's not really a blocker - it will just not take effect on those (so the impact of this PR will be limited).
But with the above, I would say that we can merge this PR now to soak it for longer.
[And we will merge the PR switching the behavior for limit soon on top of this]

@serathius @p0lyn0mial

**Pull request merged**